api/kubernetes/cli/pod.go

```diff
@@ -0,0 +1,121 @@
+// - The shell pod will be automatically removed if it's not ready after specified period of time
+// - The shell pod will be automatically removed after a specified max life (prevent zombie pods)
+// - The shell pod will be automatically removed if request is cancelled (or client closes websocket connection)
+func (kcl *KubeClient) CreateUserShellPod(ctx context.Context, serviceAccountName string) (*portainer.KubernetesShellPod, error) {
+	// Schedule the pod for automatic removal
+	maxPodKeepAlive := 1 * time.Hour
+	maxPodKeepAliveSecondsStr := fmt.Sprintf("%d", int(maxPodKeepAlive.Seconds()))
+
+	podPrefix := userShellPodPrefix(serviceAccountName)
+
+	podSpec := &v1.Pod{
+		ObjectMeta: metav1.ObjectMeta{
+			GenerateName: podPrefix,
+			Namespace:    portainerNamespace,
+			Annotations: map[string]string{
+				"kubernetes.io/pod.type": "kubectl-shell",
+			},
+		},
+		Spec: v1.PodSpec{
+			TerminationGracePeriodSeconds: new(int64),
+			ServiceAccountName:            serviceAccountName,
+			Containers: []v1.Container{
+				{
+					Name: "kubectl-shell-container",
+					// TODO update the image details
+					Image:   ShellPodImage,
+					Command: []string{"sleep"},
+					// Specify sleep time to prevent zombie pods in case portainer process is terminated
+					Args:            []string{maxPodKeepAliveSecondsStr},
+					ImagePullPolicy: v1.PullIfNotPresent,
+				},
+			},
+			RestartPolicy: v1.RestartPolicyNever,
+		},
+	}
+
+	shellPod, err := kcl.cli.CoreV1().Pods(portainerNamespace).Create(podSpec)
+	if err != nil {
+		return nil, errors.Wrap(err, "error creating shell pod")
+	}
+
+	// Wait for pod to reach ready state
+	timeoutCtx, cancelFunc := context.WithTimeout(ctx, 20*time.Second)
+	defer cancelFunc()
+	err = kcl.waitForPodStatus(timeoutCtx, v1.PodRunning, shellPod)
+	if err != nil {
+		kcl.cli.CoreV1().Pods(portainerNamespace).Delete(shellPod.Name, nil)
+		return nil, errors.Wrap(err, "aborting pod creation; error waiting for shell pod ready status")
+	}
+
+	if len(shellPod.Spec.Containers) != 1 {
+		kcl.cli.CoreV1().Pods(portainerNamespace).Delete(shellPod.Name, nil)
+		return nil, fmt.Errorf("incorrect shell pod state, expecting single container to be present")
+	}
+
+	podData := &portainer.KubernetesShellPod{
+		Namespace:        shellPod.Namespace,
+		PodName:          shellPod.Name,
+		ContainerName:    shellPod.Spec.Containers[0].Name,
+		ShellExecCommand: "/bin/bash",
+	}
+
+	// Handle pod lifecycle/cleanup - terminate pod after maxPodKeepAlive or upon request (long-lived) cancellation
+	go func() {
+		select {
+		case <-time.After(maxPodKeepAlive):
+			log.Println("[DEBUG] [internal,kubernetes/pod] [message: pod removal schedule duration exceeded]")
+			kcl.cli.CoreV1().Pods(portainerNamespace).Delete(shellPod.Name, nil)
+		case <-ctx.Done():
+			err := ctx.Err()
+			log.Printf("[DEBUG] [internal,kubernetes/pod] [message: context error: err=%s ]\n", err)
```

Comment:
Does this mean `err` is always non-nil here?

Reply:
yep; if `ctx.Done` is closed, then it must have an error.
In the context package you can find the explanation - https://golang.org/pkg/context/:

```go
    // If Done is not yet closed, Err returns nil.
    // If Done is closed, Err returns a non-nil error explaining why:
    // Canceled if the context was canceled
    // or DeadlineExceeded if the context's deadline passed.
    // After Err returns a non-nil error, successive calls to Err return the same error.
```